util/src/main/java/tc/oc/pgm/util/bukkit/Platform.java

```diff
@@ -0,0 +1,82 @@
+package tc.oc.pgm.util.bukkit;
+
+import java.lang.reflect.InvocationTargetException;
+import org.bukkit.Bukkit;
+import tc.oc.pgm.util.ClassLogger;
+import tc.oc.pgm.util.nms.NMSHacksNoOp;
+import tc.oc.pgm.util.nms.NMSHacksPlatform;
+import tc.oc.pgm.util.nms.v1_8.NMSHacks1_8;
+import tc.oc.pgm.util.nms.v1_8.NMSHacksSportPaper;
+import tc.oc.pgm.util.nms.v1_9.NMSHacks1_9;
+
+public enum Platform {
+  UNKNOWN("UNKNOWN", "UNKNOWN", NMSHacksNoOp.class, false),
+  SPORTPAPER_1_8("SportPaper", "1.8", NMSHacksSportPaper.class, false),
+  SPIGOT_1_8("Spigot", "1.8", NMSHacks1_8.class, false),
+  PAPER_1_8("Paper", "1.8", NMSHacks1_8.class, false),
+  SPIGOT_1_9("Spigot", "1.9", NMSHacks1_9.class, true),
+  PAPER_1_9("Paper", "1.9", NMSHacks1_9.class, true);
+
+  private static ClassLogger logger = ClassLogger.get(Platform.class);;
+  public static Platform SERVER_PLATFORM = computeServerPlatform();
+
+  private static Platform computeServerPlatform() {
+    String versionString = Bukkit.getServer().getVersion();
+    for (Platform platform : Platform.values()) {
+      if (versionString.contains(platform.variant)
+          && versionString.contains("MC: " + platform.majorVersion)) {
+        return platform;
+      }
+    }
+    return UNKNOWN;
+  }
+
+  private final String variant;
+  private final String majorVersion;
+  private final Class<? extends NMSHacksPlatform> nmsHacksClass;
+  private final boolean requiresProtocolLib;
+
+  Platform(
+      String variant,
+      String majorVersion,
+      Class<? extends NMSHacksPlatform> nmsHacksClass,
+      boolean requiresProtocolLib) {
+    this.variant = variant;
+    this.majorVersion = majorVersion;
+    this.nmsHacksClass = nmsHacksClass;
+    this.requiresProtocolLib = requiresProtocolLib;
+  }
+
+  public NMSHacksPlatform getNMSHacks() {
+    if (this == UNKNOWN) {
+      Bukkit.getServer().getPluginManager().disablePlugin(BukkitUtils.getPlugin());
+      throw new UnsupportedOperationException("UNKNOWN Platform!");
+    }
+    if (this.requiresProtocolLib
+        && !Bukkit.getServer().getPluginManager().isPluginEnabled("ProtocolLib")) {
+      Bukkit.getServer().getPluginManager().disablePlugin(BukkitUtils.getPlugin());
+      throw new UnsupportedOperationException(
+          "ProtocolLib is required for PGM to run on " + this.toString());
+    }
+    if (this == SERVER_PLATFORM) {
+      try {
+        logger.info("Detected server: " + this.toString());
+        return nmsHacksClass.getConstructor().newInstance();
```

Comment:
since java 8 i'm more of a fan of functions vs calling constructors with reflection.
Instead of asking for a `Class<? extends NMSHacksPlatform>` ask for a `Supplier<? extends NMSHacksPlatform>`, the supplier can be simply `NmsHacks1_8::new`, and you can call it with `supplier.get();`

Reply:
I'm fine with doing it that way. Will do

Reply:
NMSHacks1_8 caused this not work.